elastic/logstash/pipeline/logstash.conf

```diff
@@ -15,126 +17,197 @@ filter {
-      }
       mutate {
-        add_field => { "read_timestamp" => "%{@timestamp}" }
-      }
-      date {
-        match => [ "[nginx][access][time]", "dd/MMM/YYYY:H:m:s Z" ]
-        remove_field => "[nginx][access][time]"
+        split => { "[@metadata][jwt]" => "." }
       }
-      useragent {
-        source => "[nginx][access][agent]"
-        target => "[nginx][access][user_agent]"
-        remove_field => "[nginx][access][agent]"
+
+      mutate {
+        copy => { "[@metadata][jwt][1]" => "[@metadata][b64]" }
       }
-      geoip {
-        source => "[nginx][access][remote_ip]"
-        target => "[nginx][access][geoip]"
+
+      ruby {
+        init => "require 'base64'"
+        code => "event.set('[@metadata][jwt_decoded]', Base64.decode64(event.get('[@metadata][b64]')))"
       }
 
-      geoip {
-        source => "[nginx][access][client_ip]"
-        target => "[nginx][access][server][geoip]"
+      json {
+        source => "[@metadata][jwt_decoded]"
+        target => "jwt_data"
       }
+    }
 
+    }
 
-      # decode JWT
-      if [@metadata][jwt] {
+    if [nginx][access][payload] {
+      # remove \ from payload
+      mutate {
+        gsub => [
+          "[nginx][access][payload]", "[\\]", ""
+        ]
+      }
 
+      # parse json to data field
+      json {
+        source => "[nginx][access][payload]"
+        target => "[data]"
+      }
+    }
 
-        mutate {
-          split => { "[@metadata][jwt]" => "." }
-        }
 
-        mutate {
-          copy => { "[@metadata][jwt][1]" => "[@metadata][b64]" }
-        }
+    # Get the type of operation of interest
 
-        ruby {
-          init => "require 'base64'"
-          code => "event.set('[@metadata][jwt_decoded]', Base64.decode64(event.get('[@metadata][b64]')))"
-        }
+    mutate {
+      add_field => { "[@metadata][action]" => "%{[nginx][access][method]} %{[nginx][access][url]}" }
+    }
 
-        json {
-          source => "[@metadata][jwt_decoded]"
-          target => "jwt_data"
-        }
+    # Get the operation
+    translate {
+      field => "[@metadata][action]"
+      destination => "operation"
+      dictionary_path => "/app/logstash/dictionaries/action.yml"
+      fallback => "other"
+      regex => true
+    }
 
+    # Extract mine guid based on operation
+    if [operation] != "other" {
+      grok {
+        match => { "[nginx][access][url]" => [".*/api/mines/%{UUID:[mine][id]}", ".*/api/documents/mines/%{UUID:[mine][id]}"] }
       }
 
-      if [nginx][access][payload] {
-        # remove \ from payload
-        mutate {
-          gsub => [
-            "[nginx][access][payload]", "[\\]", ""
-          ]
+      if [mine][id] {
+            # lookup by mine id
+        jdbc_streaming {
+          jdbc_driver_library => "/app/logstash/jdbc/postgresql-42.2.5.jar"
+          jdbc_driver_class => "org.postgresql.Driver"
+          jdbc_connection_string => "${JDBC_CONNECTION_STRING}"
+          jdbc_user => "${JDBC_USER}"
```

Comment:
Do we need the secret JSON BC/DC set up here?

Reply:
Please provide more info about what secret JSON BC/DC is. Thanks.

Reply:
Hi! So Openshift has a BuildConfig(BC) and DeploymentConfig(DC). Environment variables are runtime configuration so Deployment Config would need to be updated to support the newly added environment variables.

Current logstash deployment config is located under `openshift/tools/logstash.dc.json` The environment variables are defined around line 108 `https://github.com/bcgov/mds/blob/develop/openshift/tools/logstash.dc.json#L108`.
You would need to add 3 more blocks alongside the rest of the environment variables for the `JDBC_CONNECTION_STRING ` , `JDBC_USER ` and the `JDBC_PASSWORD ` 

Now since the code is public, we don't hardcode the database credentials into the deployment config and they come from openshift secrets.  You would need to define the json block so that the env variable can come from secrets and openshift will take care of the rest.

You can find a similar example here:
https://github.com/bcgov/mds/blob/develop/openshift/_python36.dc.json#L454
This one is a tool called flyway which also uses JDBC to connect to postgres so I assume the environment variables . will be similar to `PGUSER `, `FLYWAY_URL ` and `PGPASSWORD `.
Please ensure for `PGUSER` you are using `database-user` and not `database-admin-user`.

After the environment variable block has been updated, you will need to update the logstash deployment config template to add the name of the secret as a parameter like done here:
https://github.com/bcgov/mds/blob/develop/openshift/_python36.dc.json#L28

Once all of that is done, you just need to pass the value in from the `pipeline/config.groovy`, this file gets picked up by the pipeline to add parameters to the deployment configs.
The logstash deployment config is here `https://github.com/bcgov/mds/blob/develop/pipeline/config.groovy#L351`
You will need to add another parameter similar to here
https://github.com/bcgov/mds/blob/develop/pipeline/config.groovy#L312

Let me know if you get stuck or have any questions

Reply:
Thank you Gyan.

Reply:
Changes have been made to dc json defintion and config.groovy for the pipeline